File: JavaGLFiguresApp/app/src/main/java/com/example/javaglfiguresapp/MyRenderer.java
package com.example.javaglfiguresapp;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.opengl.GLU;
import android.view.MotionEvent;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;


public class MyRenderer implements GLSurfaceView.Renderer {
    Context context;

    private final Square mSquare;
    private final Cube mCube;
    private final Sphere mSphere;


    private static float angleCube = 0;
    private float mAngle = 0;

    private int currentFigure = 0; // Индекс текущей фигуры: 0 - куб, 1 - сфера, 2 - прямоугольник

    public MyRenderer(Context context){
        this.context = context;
        mSquare = new Square();
        mCube = new Cube();
        mSphere = new Sphere(10);
    }

    @Override
    public void onSurfaceCreated(GL10 gl, EGLConfig config) {
        gl.glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
        gl.glClearDepthf(1.0f);
        gl.glEnable(GL10.GL_DEPTH_TEST);
        gl.glDepthFunc(GL10.GL_LEQUAL);
        gl.glHint(GL10.GL_PERSPECTIVE_CORRECTION_HINT, GL10.GL_NICEST);
        gl.glShadeModel(GL10.GL_SMOOTH);
        gl.glDisable(GL10.GL_DITHER);
    }

    @Override
    public void onSurfaceChanged(GL10 gl, int width, int height) {
        if (height == 0) height = 1;
        float aspect = (float)width / height;
        gl.glViewport(0, 0, width, height);
        gl.glMatrixMode(GL10.GL_PROJECTION);
        gl.glLoadIdentity();
        GLU.gluPerspective(gl, 45, aspect, 0.1f, 100.f);
        gl.glMatrixMode(GL10.GL_MODELVIEW); // Select model-view matrix
        gl.glLoadIdentity();
    }

    @Override
    public void onDrawFrame(GL10 gl) {
        gl.glClear(GL10.GL_COLOR_BUFFER_BIT | GL10.GL_DEPTH_BUFFER_BIT);

        gl.glLoadIdentity();
        gl.glTranslatef(0.0f, 0.0f, -5.0f);
        gl.glRotatef(mAngle, 0, 0, -1);

        // Отрисовка текущей фигуры
        switch (currentFigure) {
            case 0: // Куб
                gl.glTranslatef(0.0f, 0.0f, -2.0f);
                gl.glScalef(0.6f, 0.6f, 0.6f);
                gl.glRotatef(angleCube, 1.0f, 1.0f, 0.5f);
                mCube.draw(gl);
                break;
            case 1: // Сфера
                //gl.glTranslatef(0f, 1.0f, -4.5f);
                gl.glScalef(0.05f, 0.05f, 0.05f);
                gl.glRotatef(mAngle, 1, 1, -1);
                mSphere.draw(gl);
                break;
            case 2: // Прямоугольник
                gl.glTranslatef(0.0f, 0.0f, -5.0f);
                gl.glRotatef(mAngle, 0, 0, -1);
                mSquare.draw(gl);
                break;
        }

        float speedCube = -1.5f;
        angleCube += speedCube;
        mAngle += 1;
    }

    public boolean onTouchEvent(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            // При нажатии на экран переключаемся на следующую фигуру
            currentFigure = (currentFigure + 1) % 3; // 0, 1, 2 - куб, сфера, прямоугольник
        }
        return true;
    }
}

File: KotlinCourseProject/app/src/main/java/com/example/kotlincourseproject/MyRenderer.kt
package com.example.kotlincourseproject

import android.annotation.SuppressLint
import android.content.Context
import android.opengl.GLES20.*
import android.opengl.GLSurfaceView
import android.opengl.Matrix
import android.view.MotionEvent
import android.view.View
import com.example.kotlincourseproject.gl.GLObject
import com.example.kotlincourseproject.gl.IGLObject
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10


@SuppressLint("ClickableViewAccessibility")
class MyRenderer(ctx: Context, private val glSurfaceView: GLSurfaceView) : GLSurfaceView.Renderer,
    View.OnTouchListener {
    private val candleObjects: MutableList<IGLObject> = mutableListOf()
    private val otherObjects: MutableList<IGLObject> = mutableListOf()

    private var currentCandleIndex = 0 // Индекс текущей свечи
    private var framesPerCandle = 5 // Количество кадров для каждой свечи
    private var framesCount = 0 // Счетчик кадров

    private val bgColorR: Float = 0.3f
    private val bgColorG: Float = 0.3f
    private val bgColorB: Float = 0.3f
    private val bgColorA: Float = 1.0f


    init {
        val baseShader = Utils.inputStreamToString(ctx.assets.open("shaders/base_shader.vert"))
        val colorShader = Utils.inputStreamToString(ctx.assets.open("shaders/color_shader.frag"))
        val textureShader =
            Utils.inputStreamToString(ctx.assets.open("shaders/texture_shader.frag"))

        val table = GLObject.fromInputStream(
            baseShader,
            textureShader,
            ctx.assets.open("models/wooden_table.obj"),
            ctx.assets.open("models/wooden_table.jpeg")
        )
        table.setPosition(0f, -3.2f, -11f)
        table.setRotation(0f, 30f, 0f)
        table.setScale(0.1f, 0.1f, 0.1f)

        for (i in 0..20) {
            val candleModelNumber = i % 21
            val candleFileName = "models/candles/candle$candleModelNumber.obj"
            val textureFileName = "models/candles/textures/candles_albedo.jpg"

            val candle = GLObject.fromInputStream(
                baseShader,
                textureShader,
                ctx.assets.open(candleFileName),
                ctx.assets.open(textureFileName)
            )

            candle.setPosition(table.x + 1.2f, table.y + 1.45f, table.z + 3f)
            candle.setRotation(table.rotateX, table.rotateY, table.rotateZ)
            candle.setScale(0.6f, 0.6f, 0.6f)

            candleObjects.add(candle)
        }

        val mug = GLObject.fromInputStream(
            baseShader,
            textureShader,
            ctx.assets.open("models/mug.obj"),
            ctx.assets.open("models/mug.jpg")
        )
        mug.setPosition(table.x - 0.2f, table.y + 1.5f, table.z + 2f)
        mug.setRotation(table.rotateX, table.rotateY, table.rotateZ)
        mug.setScale(2.2f, 2.2f, 2.2f)

        val strawberry_cat = GLObject.fromInputStream(
            baseShader,
            textureShader,
            ctx.assets.open("models/strawberry_cat.obj"),
            ctx.assets.open("models/strawberry_cat.png")
        )
        strawberry_cat.setPosition(table.x + 0.4f, table.y + 1.4f, table.z + 2.5f)
        strawberry_cat.setRotation(table.rotateX, table.rotateY, table.rotateZ)
        strawberry_cat.setScale(0.15f, 0.15f, 0.15f)

        val banana_cat = GLObject.fromInputStream(
            baseShader,
            textureShader,
            ctx.assets.open("models/banana_cat.obj"),
            ctx.assets.open("models/banana_cat.png")
        )
        banana_cat.setPosition(table.x + 0.9f, table.y + 1.4f, table.z + 2.5f)
        banana_cat.setRotation(table.rotateX, table.rotateY, table.rotateZ)
        banana_cat.setScale(0.15f, 0.15f, 0.15f)

        val pumpkin = GLObject.fromInputStream(
            baseShader,
            textureShader,
            ctx.assets.open("models/pumpkin.obj"),
            ctx.assets.open("models/pumpkin.png")
        )
        pumpkin.setPosition(table.x, table.y + 1.8f, table.z - 2f)
        pumpkin.setRotation(table.rotateX, table.rotateY, table.rotateZ)
        pumpkin.setScale(1f, 1f, 1f)

        val abstract = GLObject.fromInputStream(
            baseShader,
            textureShader,
            ctx.assets.open("models/abstract.obj"),
            ctx.assets.open("models/abstract.png")
        )
        abstract.setPosition(table.x - 1.6f, table.y + 2.2f, table.z - 2f)
        abstract.setRotation(table.rotateX, table.rotateY, table.rotateZ)
        abstract.setScale(1f, 1f, 1f)

        otherObjects.add(table)
        otherObjects.add(mug)
        otherObjects.add(strawberry_cat)
        otherObjects.add(banana_cat)
        otherObjects.add(pumpkin)
        otherObjects.add(abstract)

        glSurfaceView.setOnTouchListener(this)
    }

    override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
        glEnable(GL_DEPTH_TEST)
        glDepthFunc(GL_LEQUAL)

        glEnable(GL_BLEND)
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)

        for (obj in candleObjects + otherObjects) {
            obj.onSurfaceCreated(gl, config)
        }
    }

    override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
        glViewport(0, 0, width, height)
        for (obj in candleObjects + otherObjects) {
            obj.onSurfaceChanged(gl, width, height)
        }
    }

//    override fun onDrawFrame(gl: GL10) {
//        glClearColor(bgColorR, bgColorG, bgColorB, bgColorA)
//        glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)
//
//        drawCandles(gl)
//        drawOtherObjects(gl)
//    }

    private fun drawCandles(gl: GL10) {
        val currentCandle = (candleObjects[currentCandleIndex] as GLObject)

        currentCandle.setLightDirection(0f, 1f, 0f)
        currentCandle.onDrawFrame(gl)

        framesCount++
        if (framesCount >= framesPerCandle) {
            framesCount = 0
            currentCandleIndex = (currentCandleIndex + 1) % candleObjects.size
        }
    }

    private fun drawOtherObjects(gl: GL10) {
        val currentCandle = (candleObjects[currentCandleIndex] as GLObject)

        for (obj in otherObjects) {
            obj.setLightDirection(currentCandle.x - obj.x, currentCandle.y + 1f - obj.y, currentCandle.z - obj.z)
            obj.onDrawFrame(gl)
        }
    }

    override fun onTouch(view: View?, event: MotionEvent?): Boolean {
        when (event?.action) {
            MotionEvent.ACTION_DOWN -> {
                startRotation()
            }
            MotionEvent.ACTION_UP -> {
                stopRotation()
            }
        }
        return true
    }

    private var isRotating = false

    private fun startRotation() {
        isRotating = true
    }

    private fun stopRotation() {
        isRotating = false
        resetRotation()
    }

    private fun resetRotation() {
        for (obj in candleObjects + otherObjects) {
            //obj.setRotation(0f, 30f, 0f)
        }
    }

    override fun onDrawFrame(gl: GL10) {
        glClearColor(bgColorR, bgColorG, bgColorB, bgColorA)
        glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)

        if (isRotating) {
            for (obj in otherObjects) {
                if (obj !is GLObject || obj !== otherObjects[0]) {
                    obj.rotateY += 2.5f
                }
            }
        }

        drawCandles(gl)
        drawOtherObjects(gl)
    }
}

